gtsam/nonlinear/CustomFactor.cpp

```diff
@@ -0,0 +1,35 @@
+/* ----------------------------------------------------------------------------
+
+ * GTSAM Copyright 2010, Georgia Tech Research Corporation,
+ * Atlanta, Georgia 30332-0415
+ * All Rights Reserved
+ * Authors: Frank Dellaert, et al. (see THANKS for the full author list)
+
+ * See LICENSE for the license information
+
+ * -------------------------------------------------------------------------- */
+
+/**
+ * @file    CustomFactor.cpp
+ * @brief   Class to enable arbitrary factors with runtime swappable error function.
+ * @author  Fan Jiang
+ */
+
+#include <gtsam/nonlinear/CustomFactor.h>
+
+namespace gtsam {
+
+Vector CustomFactor::unwhitenedError(const Values& x, boost::optional<std::vector<Matrix>&> H) const {
+  if(this->active(x)) {
+    if(H) {
+      return this->errorFunction(*this, x, H.get_ptr());
+    } else {
+      JacobianVector dummy;
+      return this->errorFunction(*this, x, &dummy);
```

Comment:
could this pass in null_ptr instead or something?  Possibly by something along the lines of

```c++
typedef std::function<Vector(const CustomFactor&, const Values&, std::optional<JacobianVector*>)> CustomErrorFunction;
```
or something ???

Reply:
That's apparently doable as per some issues on Pybind11, but we'd have to use boost::optional because cpp11

Reply:
Yes, done :)